akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -657,40 +657,58 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    */
   @InternalApi private[stream] var lastCancellationCause: Throwable = _
 
+  private lazy val delayCancellationCallback =
+    getAsyncCallback[(Throwable, Attributes.CancellationStrategy.Strategy)] {
+      case (cause, strategy) => internalCancelStage(cause, strategy)
+    }
+
   /**
    * Automatically invokes [[cancel()]] or [[complete()]] on all the input or output ports that have been called,
    * then marks the stage as stopped.
    */
-  final def cancelStage(cause: Throwable): Unit = {
-    // TODO: It's debatable if completing the stage if one output is cancelled is the right way to do things.
-    //       At least optionally it might be more reasonable to fail the stage with the given cause. That
-    //       would mean that all other *outputs* are failed, i.e. it would only concern stages with more that one
-    //       output anyway.
-    var i = 0
-    while (i < portToConn.length) {
-      if (i < inCount)
-        interpreter.cancel(portToConn(i), cause)
-      else
-        handlers(i) match {
-          case e: Emitting[_] => e.addFollowUp(new EmittingCompletion(e.out, e.previous))
-          case _              => interpreter.complete(portToConn(i))
+  final def cancelStage(cause: Throwable): Unit =
+    internalCancelStage(
+      cause,
+      attributes.get[Attributes.CancellationStrategy](Attributes.CancellationStrategy.Default).strategy)
+
+  private def internalCancelStage(cause: Throwable, strategy: Attributes.CancellationStrategy.Strategy): Unit = {
+    import Attributes.CancellationStrategy._
+    import SubscriptionWithCancelException._
+    strategy match {
+      case CompleteStage => internalCompleteStage(cause, OptionVal.None)
+      case FailStage     => internalCompleteStage(cause, OptionVal.Some(cause))
+      case PropagateFailure =>
+        cause match {
+          case NoMoreElementsNeeded | StageWasCompleted => internalCompleteStage(cause, OptionVal.None)
+          case _                                        => internalCompleteStage(cause, OptionVal.Some(cause))
         }
-      i += 1
+      case AfterDelay(delay, andThen) =>
+        materializer.scheduleOnce(delay, () => delayCancellationCallback.invoke((cause, andThen)))
```

Comment:
I have a hard time understanding the consequences of this one wrt this stage being kind of failed but still running, will upstream potentially still push elements?

Reply:
Mmh, that's a good point. We should make sure that we ignore any upstream elements until the cancellation is finally done. Completion and failures should come through, though (that's basically the purpose of the whole thing).

I'll add a test and maybe some code.

Reply:
This is indeed a problem. The stage called `cancelStage`, explicitly or implicitly through the default implementation of `onDownstreamFinish`, but then it still isn't stopped and may continue to process events. We only want that to the degree that it propagates errors.

Reply:
Would the interpreter need to pause it somehow?

Reply:
I'm now experimenting with shifting the cancellation delay handling to the downstream side of the cancelled connection. That way the receiver of the cancellation would not even know yet that it will be cancelled in a while.

Reply:
That seems to work, see the update. The downside is that the handling is now not as simple as before. 

`AfterDelay` is now handled in the stage that calls `cancel(in, cause)` and the other strategies are handled in the stage that are receiving the cancellation (i.e. where `cancelStage` is called).